lib/kdl/include/kdl/expected.h

```diff
@@ -0,0 +1,627 @@
+         *
+         * @tparam Visitor the type of the visitor
+         * @param visitor the visitor to apply
+         * @param expected the expectation to apply the visitor to
+         * @return the result of applying the given visitor
+         */
+        template <typename Visitor>
+        friend auto visit(Visitor&& visitor, expected&& expected) {
+            return absl::visit(
+                std::forward<Visitor>(visitor),
+                std::move(expected.m_value));
+        }
+        
+        /**
+         * Applies the given function to the contained value and returns a new expectation with the result type of the
+         * given function as its value type.
+         *
+         * If this expectation contains a value, the function is applied to it and the result of applying the function
+         * is returned wrapped in an expectation.
+         * If this expectation contains an error, that error is returned as is.
+         *
+         * The value is passed to the mapping function by const lvalue reference.
+
+         * @tparam F the type of the function to apply
+         * @param f the function to apply
+         */
+        template <typename F>
+        auto map(F&& f) const & {
+            using R = std::invoke_result_t<F, Value>;
+            return visit(kdl::overload {
+                [&](const value_type& v) { return expected<R, Errors...>(f(v)); },
+                [] (const auto& e)       { return expected<R, Errors...>(e); }
+            }, *this);
+        }
+        
+        /**
+         * Applies the given function to the contained value and returns a new expectation with the result type of the
+         * given function as its value type.
+         *
+         * If this expectation contains a value, the function is applied to it and the result of applying the function
+         * is returned wrapped in an expectation.
+         * If this expectation contains an error, that error is returned as is.
+         *
+         * The value is passed to the mapping function by rvalue reference.
+         *
+         * The given function can move the value out of this expectation, so care must be taken not to
+         * use this expectation afterwards, as it will be in a moved-from state.
+
+         * @tparam F the type of the function to apply
+         * @param f the function to apply
+         */
+        template <typename F>
+        auto map(F&& f) && {
+            using R = std::invoke_result_t<F, Value>;
+            return visit(kdl::overload {
+                [&](value_type&& v) { return expected<R, Errors...>(f(std::move(v))); },
+                [] (const auto& e)  { return expected<R, Errors...>(e); }
+            }, *this);
+        }
+
+        /**
+         * Indicates whether this expectation contains a value.
+         */
+        bool success() const {
+            return absl::holds_alternative<value_type>(m_value);
```

Comment:
Note, from https://en.cppreference.com/w/cpp/utility/variant/holds_alternative , "The call is ill-formed if T does not appear exactly once in Types..."

This could be written as `m_value.index()` == 0 to not have that restriction, but I guess other code does.

Reply:
You mean for the case when the value type is also used as an error type? Good point.

Reply:
Yeah

Reply:
done